google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/application/GoogleApiClientAppEngineAdminService.java

```diff
@@ -40,18 +47,62 @@
 
   private static final String APP_ENGINE_RESOURCE_WILDCARD = "-";
   private static final long CREATE_APPLICATION_POLLING_INTERVAL_MS = 1000;
+  private static final int APPLICATION_CACHE_MAX_SIZE = 10000;
+  private final long LOCATION_CACHE_REFRESH_PERIOD_MILLIS = 1000 * 60 * 60 * 24; // 1 day
+
+  // Cache of GCP application resources. This assumes that project IDs are globally unique. Null or
+  // missing applications are not cached because they cannot be reliably invalidated.
+  private final Cache<String, Application> appEngineApplicationCache = CacheBuilder.newBuilder()
+      // arbitrary size limit
+      .maximumSize(APPLICATION_CACHE_MAX_SIZE)
+      // Even though these values should never change, it won't kill us to refresh once per day.
+      .expireAfterWrite(1, TimeUnit.DAYS)
+      .build();
+
+  // Cache of all available GCP locations. This list is expected to change very infrequently.
+  private final List<Location> appEngineLocationCache = new CopyOnWriteArrayList<>();
```

Comment:
That's a good point; that would work. I played around with that approach, and while it does lead to cleaner code than we currently have here, I still vote against it, since by not requiring the user to call the `getAllRegions` method with a valid credential, we open ourselves up to a new class of issues, i.e. if that method is called without a logged in user. I really don't want to depend on that external state in order to have this call succed. 

I think the best compromise is to use a map-based cache with a single value, as that would support time-based eviction, and allows us to pass credentials at lookup time.